packages/select/src/Select.tsx

```diff
@@ -0,0 +1,488 @@
+  // match the current state and use it instead.
+  useEffect(() => {
+    if (uncontrolledSelectedOption !== null) {
+      setUncontrolledSelectedOption(
+        reconcileOption(children, uncontrolledSelectedOption) ??
+          initialUncontrolledSelectedOption,
+      );
+    }
+  }, [children, initialUncontrolledSelectedOption, uncontrolledSelectedOption]);
+
+  const selectedOption = useMemo(() => {
+    if (value !== undefined) {
+      let selectedOption: OptionElement | null = null;
+
+      traverseSelectChildren(children, (option, group) => {
+        if (isOptionSelectable(option, group, value)) {
+          selectedOption = option;
+        }
+      });
+      return selectedOption;
+    } else {
+      return uncontrolledSelectedOption;
+    }
+  }, [children, uncontrolledSelectedOption, value]);
+
+  const onSelect = useCallback(
+    (option: OptionElement | null) => {
+      if (value === undefined) {
+        setUncontrolledSelectedOption(option);
+      }
+      onChange?.(getOptionValue(option));
+      setFocusedOption(undefined);
+      onClose();
+    },
+    [onChange, onClose, value],
+  );
+
+  const getOptionClickHandler = useCallback(
+    (option: OptionElement | null, optionDisabled: boolean) => {
+      return (event: React.MouseEvent) => {
+        event.preventDefault();
+        event.stopPropagation();
+
+        if (!disabled && !optionDisabled) {
+          onSelect(option);
+          onClose();
+        }
+      };
+    },
+    [disabled, onClose, onSelect],
+  );
+
+  const onDeselect = useCallback(() => {
+    onSelect(null);
+  }, [onSelect]);
+
+  /**
+   * Focus
+   */
+
+  const [focusedOption, setFocusedOption] = useState<
+    OptionElement | null | undefined
+  >();
+
+  const enabledOptions = useMemo(() => {
+    const enabledOptions: Array<OptionElement | null> = [null];
+
+    traverseSelectChildren(children, (option, group) => {
+      if (!isOptionDisabled(option, group)) {
+        enabledOptions.push(option);
+      }
+    });
+
+    return enabledOptions;
+  }, [children]);
+
+  const onSelectFocusedOption = useCallback(() => {
+    if (focusedOption !== undefined) {
+      onSelect(focusedOption);
+    }
+  }, [focusedOption, onSelect]);
+
+  const onFocusFirstOption = useCallback(() => {
+    setFocusedOption(null);
+  }, []);
+
+  const onFocusLastOption = useCallback(() => {
+    setFocusedOption(enabledOptions[enabledOptions.length - 1]);
+  }, [enabledOptions]);
+
+  const onFocusPreviousOption = useCallback(() => {
+    if (focusedOption === undefined) {
+      onFocusLastOption();
+    } else {
+      const index = Math.max(0, enabledOptions.indexOf(focusedOption) - 1);
+      setFocusedOption(enabledOptions[index]);
+    }
+  }, [enabledOptions, focusedOption, onFocusLastOption]);
+
+  const onFocusNextOption = useCallback(() => {
+    if (focusedOption === undefined) {
+      onFocusFirstOption();
+    } else {
+      const index = Math.min(
```

Comment:
We should have this wrap such that the first item gets focused after we've reached the end of the list, rather than not moving forward